src/ObjectUtils/CoffFile.cpp

```diff
@@ -258,34 +333,135 @@ ErrorMessageOr<ModuleSymbols> CoffFileImpl::LoadDebugSymbols() {
+    return ErrorMessage{"Unable to read Exception Table."};
+  }
+
+  if (exception_table_bytes.size() % sizeof(llvm::Win64EH::RuntimeFunction) != 0) {
+    return ErrorMessage{"Unable to read Exception Table: Unexpected size."};
+  }
+
+  return llvm::ArrayRef<llvm::Win64EH::RuntimeFunction>{
+      reinterpret_cast<const llvm::Win64EH::RuntimeFunction*>(exception_table_bytes.data()),
+      exception_table_bytes.size() / sizeof(llvm::Win64EH::RuntimeFunction)};
+}
+
+ErrorMessageOr<const llvm::Win64EH::RuntimeFunction*>
+orbit_object_utils::CoffFileImpl::GetPrimaryRuntimeFunction(
+    const llvm::Win64EH::RuntimeFunction* runtime_function) {
+  const llvm::Win64EH::UnwindInfo* unwind_info{};
+  while (true) {
+    uintptr_t unwind_info_bytes{};
+    if (object_file_->getRvaPtr(runtime_function->UnwindInfoOffset, unwind_info_bytes)) {
+      return ErrorMessage{absl::StrFormat("Unable to read RUNTIME_FUNCTION at RVA %#x.",
+                                          runtime_function->UnwindInfoOffset)};
+    }
+
+    unwind_info = reinterpret_cast<const llvm::Win64EH::UnwindInfo*>(unwind_info_bytes);
+    // From https://docs.microsoft.com/en-us/cpp/build/exception-handling-x64#struct-unwind_info:
+    // "UNW_FLAG_CHAININFO: This unwind info structure is not the primary one for the procedure.
+    // Instead, the chained unwind info entry is the contents of a previous RUNTIME_FUNCTION entry."
+    if ((unwind_info->getFlags() & llvm::Win64EH::UNW_ChainInfo) == 0) {
+      break;
+    }
+
+    runtime_function = unwind_info->getChainedFunctionEntry();
+  }
+
+  return runtime_function;
+}
+
+ErrorMessageOr<std::vector<CoffFileImpl::UnwindRange>> CoffFileImpl::GetUnwindRanges() {
+  auto runtime_functions_or_error = GetRuntimeFunctions();
+  if (runtime_functions_or_error.has_error()) {
+    return ErrorMessage{absl::StrFormat("Unable to load unwind info ranges: %s",
+                                        runtime_functions_or_error.error().message())};
+  }
+  const llvm::ArrayRef<llvm::Win64EH::RuntimeFunction>& runtime_functions =
+      runtime_functions_or_error.value();
+
+  std::vector<UnwindRange> unwind_ranges;
+  // For a function, the "primary" RUNTIME_FUNCTION and the ones with UNW_FLAG_CHAININFO set (if
+  // any) form a tree rooted on the "primary" RUNTIME_FUNCTION. We merge all such RUNTIME_FUNCTIONs
+  // from the same tree into a single UnwindRange.
+  for (const llvm::Win64EH::RuntimeFunction& runtime_function : runtime_functions) {
+    auto primary_runtime_function_or_error = GetPrimaryRuntimeFunction(&runtime_function);
+    if (primary_runtime_function_or_error.has_error()) {
+      return ErrorMessage{absl::StrFormat("Unable to load unwind info ranges: %s",
+                                          primary_runtime_function_or_error.error().message())};
+    }
+    const llvm::Win64EH::RuntimeFunction* primary_runtime_function =
+        primary_runtime_function_or_error.value();
+
+    const uint64_t primary_address =
```

Comment:
I realized I had one more question here: Do we know that the start address of the primary address is always <= than the end address of the runtime function?

Reply:
I think so. https://docs.microsoft.com/en-us/cpp/build/exception-handling-x64?view=msvc-170#struct-unwind_info states "UNW_FLAG_CHAININFO: This unwind info structure is not the primary one for the procedure. Instead, the chained unwind info entry is the contents of a **previous** RUNTIME_FUNCTION entry". And RUNTIME_FUNCTIONs are sorted by address.

Reply:
Ah, perfect. Thanks for clarifying!